### PR TITLE
:computer: Identify interface nodes with multiple material contact

### PR DESCRIPTION
**Pull request by kks32**
Identify interface nodes

### CODE REVIEW

File: tests/particle_test.cc [hunk 1]
```diff
@@ -1747,6 +1757,17 @@ TEST_CASE("Particle is checked for 3D case", "[particle][3D]") {
...
+    // Map particle material id to nodes
+    particle->append_material_id_to_nodes();
+    REQUIRE(node0->material_ids()[0] == 0);
+    REQUIRE(node1->material_ids()[0] == 0);
+    REQUIRE(node2->material_ids()[0] == 0);
+    REQUIRE(node3->material_ids()[0] == 0);
+    REQUIRE(node4->material_ids()[0] == 0);
+    REQUIRE(node5->material_ids()[0] == 0);
+    REQUIRE(node6->material_ids()[0] == 0);
+    REQUIRE(node7->material_ids()[0] == 0);
```
Comment (kks32) [inline tests/particle_test.cc]: Check `remove_duplicates`

File: include/cell.tcc [hunk 2]
```diff
@@ -1015,3 +1015,12 @@ template <unsigned Tdim>
 inline unsigned mpm::Cell<Tdim>::rank() const {
   return this->rank_;
 }
+
+//! Assign materials to nodes from the particles within this cell
+template <unsigned Tdim>
+inline void mpm::Cell<Tdim>::append_material_id_to_nodes(unsigned material_id) {
+  // Loop over all nodes to add the material_id to the nodes
+  for (auto itr_node = this->nodes_.begin(); itr_node != this->nodes_.end();
```
Comment (kks32) [inline include/cell.tcc]: Use a range-based for loop

File: include/mpm_explicit.tcc [hunk 3]
```diff
@@ -142,6 +142,16 @@ bool mpm::MPMExplicit<Tdim>::solve() {
 
     task_group.wait();
 
+    // Assign material ids to node
```
Comment (kks32) [inline include/mpm_explicit.tcc:148]: Add an `if` condition for only when there is an interface or at least two active material models?

File: include/node.tcc [hunk 4]
```diff
@@ -510,3 +511,19 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_friction_constraints(double dt) {
...
+  material_ids_.emplace_back(id);
+}
+
+//! Remove duplicate material ids from material_ids_
+template <unsigned Tdim, unsigned Tdof, unsigned Tnphases>
+void mpm::Node<Tdim, Tdof, Tnphases>::remove_duplicate_material_ids() {
+  // sort the vector, get the unique values and remove the duplicates
+  std::sort(material_ids_.begin(), material_ids_.end());
+  auto last = std::unique(material_ids_.begin(), material_ids_.end());
+  material_ids_.erase(last, material_ids_.end());
```
Comment (kks32) [inline include/node.tcc]:
```suggestion
  material_ids_.erase(std::unique(material_ids_.begin(), material_ids_.end(), material_ids_.end());
```

File: tests/cell_test.cc [hunk 5]
```diff
@@ -489,6 +489,27 @@ TEST_CASE("Cell is checked for 2D case", "[cell][2D]") {
...
+    auto cell = std::make_shared<mpm::Cell<Dim>>(0, Nnodes, element);
+
+    cell->add_node(0, node0);
+    cell->add_node(1, node1);
+    cell->add_node(2, node2);
+    cell->add_node(3, node3);
+
+    cell->append_material_id_to_nodes(1);
+    cell->append_material_id_to_nodes(3);
+    REQUIRE(node0->material_ids()[0] == 1);
```
Comment (kks32) [inline tests/cell_test.cc]: Check size before and after `remove_duplicates` call

File: tests/node_test.cc [hunk 6]
```diff
@@ -442,6 +442,25 @@ TEST_CASE("Node is checked for 1D case", "[node][1D]") {
...
+      node->append_material_id(2);
+      node->append_material_id(0);
+      node->append_material_id(2);
+      REQUIRE(node->material_ids()[0] == 0);
+      REQUIRE(node->material_ids()[1] == 1);
+      REQUIRE(node->material_ids()[2] == 2);
+      REQUIRE(node->material_ids()[3] == 0);
+      REQUIRE(node->material_ids()[4] == 2);
+
+      // Check removal of duplicates
```
Comment (kks32) [inline tests/node_test.cc]: Check size

Comment (kks32) [inline tests/node_test.cc]: You need to call `remove_duplicates()`

File: tests/node_test.cc [hunk 7]
```diff
@@ -1000,6 +1019,26 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
...
+      node->append_material_id(0);
+
+      node->append_material_id(2);
+      REQUIRE(node->material_ids()[0] == 0);
+      REQUIRE(node->material_ids()[1] == 1);
+      REQUIRE(node->material_ids()[2] == 2);
+      REQUIRE(node->material_ids()[3] == 0);
+      REQUIRE(node->material_ids()[4] == 2);
+
+      // Check removal of duplicates
```
Comment (kks32) [inline tests/node_test.cc]: You need to call `remove_duplicates`

File: tests/node_test.cc [hunk 8]
```diff
@@ -1534,5 +1573,24 @@ TEST_CASE("Node is checked for 3D case", "[node][3D]") {
...
+      node->append_material_id(2);
+      node->append_material_id(0);
+      node->append_material_id(2);
+      REQUIRE(node->material_ids()[0] == 0);
+      REQUIRE(node->material_ids()[1] == 1);
+      REQUIRE(node->material_ids()[2] == 2);
+      REQUIRE(node->material_ids()[3] == 0);
+      REQUIRE(node->material_ids()[4] == 2);
+
+      // Check removal of duplicates
```
Comment (kks32) [inline tests/node_test.cc]: Needs `remove_duplicates()`

File: include/particle.tcc [hunk 9]
```diff
@@ -855,3 +855,9 @@ template <unsigned Tdim>
 Eigen::VectorXd mpm::Particle<Tdim>::vector_data(const std::string& property) {
   return this->properties_.at(property)();
 }
+
+//! Assign material id of this particle to nodes
+template <unsigned Tdim>
+void mpm::Particle<Tdim>::append_material_id_to_nodes() const {
+  cell_->append_material_id_to_nodes(material_->id());
```
Comment (kks32) [inline include/particle.tcc]:
```suggestion
  cell_->append_material_id_to_nodes(material_id_);
```

File: include/mpm_explicit.tcc [hunk 10]
```diff
@@ -142,6 +142,16 @@ bool mpm::MPMExplicit<Tdim>::solve() {
...
+    mesh_->iterate_over_particles(
+        std::bind(&mpm::ParticleBase<Tdim>::append_material_id_to_nodes,
+                  std::placeholders::_1));
+
+    // Delete duplicate material ids in nodes
```
Comment (kks32) [inline include/mpm_explicit.tcc]: Remove this call

File: include/node.h [hunk 11]
```diff
@@ -205,6 +205,16 @@ class Node : public NodeBase<Tdim> {
     generic_boundary_constraints_ = true;
   }
 
+  //! Add material id from material points to list of materials in materials_
+  //! \param[in] id Material id to be stored at the node
+  void append_material_id(unsigned id) override;
+
+  //! Remove duplicate material ids from material_ids_
```
Comment (kks32) [inline include/node.h]: Delete this function

File: include/node.h [hunk 12]
```diff
@@ -205,6 +205,16 @@ class Node : public NodeBase<Tdim> {
...
+  void remove_duplicate_material_ids() override;
+
+  //! Return material ids in node
+  std::vector<unsigned> material_ids() const override { return material_ids_; }
```
Comment (kks32) [inline include/node.h]:
```suggestion
  std::set<unsigned> material_ids() const override { return material_ids_; }
```

File: include/node.h [hunk 13]
```diff
@@ -236,6 +246,8 @@ class Node : public NodeBase<Tdim> {
   std::map<unsigned, double> velocity_constraints_;
   //! Rotation matrix for general velocity constraints
   Eigen::Matrix<double, Tdim, Tdim> rotation_matrix_;
+  //! Material ids whose information was passed to this node
+  std::vector<unsigned> material_ids_;
```
Comment (kks32) [inline include/node.h]:
```suggestion
  std::set<unsigned> material_ids_;
```

File: include/node.tcc [hunk 14]
```diff
@@ -510,3 +511,19 @@ void mpm::Node<Tdim, Tdof, Tnphases>::apply_friction_constraints(double dt) {
...
 }
+
+//! Add material id from material points to material_ids_
+template <unsigned Tdim, unsigned Tdof, unsigned Tnphases>
+void mpm::Node<Tdim, Tdof, Tnphases>::append_material_id(unsigned id) {
+  std::lock_guard<std::mutex> guard(node_mutex_);
+  material_ids_.emplace_back(id);
+}
+
+//! Remove duplicate material ids from material_ids_
```
Comment (kks32) [inline include/node.tcc]: Remove this function

**Review by kks32 — COMMENTED** [review]
Use `std::set` instead of a vector

File: include/mpm_explicit.tcc [hunk 15]
```diff
@@ -142,6 +146,12 @@ bool mpm::MPMExplicit<Tdim>::solve() {
 
     task_group.wait();
 
+    // Assign material ids to node
+    if (interface_ == true)
```
Comment (kks32) [inline include/mpm_explicit.tcc]:
```suggestion
    if (interface_)
```

File: tests/cell_test.cc [hunk 16]
```diff
@@ -489,6 +489,49 @@ TEST_CASE("Cell is checked for 2D case", "[cell][2D]") {
...
+
+    // check amount of material ids added to each node
+    REQUIRE(node0->material_ids().size() == 2);
+    REQUIRE(node1->material_ids().size() == 2);
+    REQUIRE(node2->material_ids().size() == 2);
+    REQUIRE(node3->material_ids().size() == 2);
+
+    // check which material ids were added to each node
+    int id_check = 1;
+    for (auto id_itr : node0->material_ids()) {
```
Comment (kks32) [inline tests/cell_test.cc]: Check for the size, and maybe just set a vector to iterate over, this check is confusing

File: tests/node_test.cc [hunk 17]
```diff
@@ -1000,6 +1019,25 @@ TEST_CASE("Node is checked for 2D case", "[node][2D]") {
...
+      node->append_material_id(3);
+      node->append_material_id(0);
+      node->append_material_id(2);
+
+      // Check size of material_ids
+      REQUIRE(node->material_ids().size() == 4);
+
+      // Check elements of material_ids
+      int id_check = 0;
+      for (auto id_itr : node->material_ids()) {
```
Comment (kks32) [inline tests/node_test.cc]: check against a set of material ids

File: tests/cell_test.cc [hunk 18]
```diff
@@ -1,3 +1,4 @@
+#include <iostream>
```
Comment (kks32) [inline tests/cell_test.cc]: Remove this please

File: tests/cell_test.cc [hunk 19]
```diff
@@ -489,6 +490,46 @@ TEST_CASE("Cell is checked for 2D case", "[cell][2D]") {
...
+
+    // check amount of material ids added to each node
+    REQUIRE(node0->material_ids().size() == 2);
+    REQUIRE(node1->material_ids().size() == 2);
+    REQUIRE(node2->material_ids().size() == 2);
+    REQUIRE(node3->material_ids().size() == 2);
+
+    // check which material ids were added to each node
+    std::vector<unsigned> material_ids = {1, 2};
+    unsigned i = 0;
```
Comment (kks32) [inline tests/cell_test.cc]: Could we create a vector of nodes, and then iterate over them instead?

File: tests/cell_test.cc [hunk 20]
```diff
@@ -1562,6 +1603,70 @@ TEST_CASE("Cell is checked for 3D case", "[cell][3D]") {
...
+    REQUIRE(node2->material_ids().size() == 2);
+    REQUIRE(node3->material_ids().size() == 2);
+    REQUIRE(node4->material_ids().size() == 2);
+    REQUIRE(node5->material_ids().size() == 2);
+    REQUIRE(node6->material_ids().size() == 2);
+    REQUIRE(node7->material_ids().size() == 2);
+
+    // check which material ids were added to each node
+    std::vector<unsigned> material_ids = {1, 2};
+    unsigned i = 0;
```
Comment (kks32) [inline tests/cell_test.cc]: Could we create a vector of nodes, and then iterate over them instead?

File: tests/particle_test.cc [hunk 21]
```diff
@@ -1300,6 +1300,109 @@ TEST_CASE("Particle is checked for 2D case", "[particle][2D]") {
...
+    particle2->append_material_id_to_nodes();
+
+    // Check size of material_ids set at each node
+    REQUIRE(node0->material_ids().size() == 2);
+    REQUIRE(node1->material_ids().size() == 2);
+    REQUIRE(node2->material_ids().size() == 2);
+    REQUIRE(node3->material_ids().size() == 2);
+
+    // Check the material ids appended to nodes
+    std::vector<unsigned> material_ids = {0, 1};
```
Comment (kks32) [inline tests/particle_test.cc:1382]: Could we create a vector of nodes, and then iterate over them instead?

File: tests/particle_test.cc [hunk 22]
```diff
@@ -2346,4 +2452,147 @@ TEST_CASE("Particle is checked for 3D case", "[particle][3D]") {
...
+    REQUIRE(node0->material_ids().size() == 2);
+    REQUIRE(node1->material_ids().size() == 2);
+    REQUIRE(node2->material_ids().size() == 2);
+    REQUIRE(node3->material_ids().size() == 2);
+    REQUIRE(node4->material_ids().size() == 2);
+    REQUIRE(node5->material_ids().size() == 2);
+    REQUIRE(node6->material_ids().size() == 2);
+    REQUIRE(node7->material_ids().size() == 2);
+
+    // Check the material ids appended to nodes
```
Comment (kks32) [inline tests/particle_test.cc]: Could we create a vector of nodes, and then iterate over them instead?

File: tests/cell_test.cc [hunk 23]
```diff
@@ -492,42 +491,31 @@ TEST_CASE("Cell is checked for 2D case", "[cell][2D]") {
...
   SECTION("Check material id addition to nodes") {
+    // create cell
     auto cell = std::make_shared<mpm::Cell<Dim>>(0, Nnodes, element);
-    // add nodes to cell
-    cell->add_node(0, node0);
-    cell->add_node(1, node1);
-    cell->add_node(2, node2);
-    cell->add_node(3, node3);
+    // create a vector of nodes and add them to cell
+    std::vector<std::shared_ptr<mpm::NodeBase<Dim>>> nodes;
```
Comment (kks32) [inline tests/cell_test.cc]: std::vector<std::shared_ptr<mpm::NodeBase<Dim>>> nodes = {node0, node1, node2, node3}

File: tests/cell_test.cc [hunk 24]
```diff
@@ -492,42 +491,31 @@ TEST_CASE("Cell is checked for 2D case", "[cell][2D]") {
...
-      REQUIRE(*itr == material_ids.at(i));
-    i = 0;
-    mat_ids = node2->material_ids();
-    for (auto itr = mat_ids.begin(); itr != mat_ids.end(); ++itr, ++i)
-      REQUIRE(*itr == material_ids.at(i));
-    i = 0;
-    mat_ids = node3->material_ids();
-    for (auto itr = mat_ids.begin(); itr != mat_ids.end(); ++itr, ++i)
-      REQUIRE(*itr == material_ids.at(i));
+    for (int j = 0; j < nodes.size(); ++j) {
```
Comment (kks32) [inline tests/cell_test.cc]: Use a range based for loop

File: tests/cell_test.cc [hunk 25]
```diff
@@ -509,10 +506,10 @@ TEST_CASE("Cell is checked for 2D case", "[cell][2D]") {
     // check if the correct amount of material ids were added to node and if
     // their indexes are correct
     std::vector<unsigned> material_ids = {1, 2};
-    for (int j = 0; j < nodes.size(); ++j) {
-      REQUIRE(nodes[j]->material_ids().size() == 2);
+    for (auto node_itr : nodes) {
```
Comment (kks32) [inline tests/cell_test.cc]: This is not an iterator it's the node so please use `for (const auto& node : nodes) {`

File: tests/cell_test.cc [hunk 26]
```diff
@@ -1615,10 +1605,10 @@ TEST_CASE("Cell is checked for 3D case", "[cell][3D]") {
...
```
Comment (kks32) [inline tests/cell_test.cc]: This is not an iterator it's the node so please use for (const auto& node : nodes) {

File: tests/particle_test.cc [hunk 27]
```diff
@@ -1383,10 +1380,10 @@ TEST_CASE("Particle is checked for 2D case", "[particle][2D]") {
     // check if the correct amount of material ids were added to node and if
     // their indexes are correct
     std::vector<unsigned> material_ids = {0, 1};
-    for (int j = 0; j < nodes.size(); ++j) {
-      REQUIRE(nodes[j]->material_ids().size() == 2);
+    for (auto node_itr : nodes) {
```
Comment (kks32) [inline tests/particle_test.cc]: This is not an iterator it's the node so please use for (const auto& node : nodes) {

File: tests/particle_test.cc [hunk 28]
```diff
@@ -2543,10 +2533,10 @@ TEST_CASE("Particle is checked for 3D case", "[particle][3D]") {
...
```
Comment (kks32) [inline tests/particle_test.cc]: This is not an iterator it's the node so please use for (const auto& node : nodes) {

**Review by cbgeo — APPROVED** [review]

**Pull request merged**